### PR TITLE
fix migrations and storage layer

### PR DESCRIPTION
**Pull request by gitsimon**

### CODE REVIEW

File: pallets/acurast/src/migration.rs [hunk 1]
```diff
@@ -29,29 +32,47 @@ pub mod v1 {
     }
 }
 
-pub fn migrate_to_v2<T: Config>() -> Weight {
+pub fn migrate<T: Config>() -> Weight {
+    let migrations: [&dyn Fn() -> Weight; STORAGE_VERSION_ - 1] =
```
Comment (RomarQ) [inline pallets/acurast/src/migration.rs:36]: Should be:
```suggestion
    let onchain_version = Pallet::<T>::on_chain_storage_version();
    let migrations: [&dyn Fn() -> Weight; onchain_version] =
```

Do you agree?

File: pallets/acurast/src/migration.rs [hunk 2]
```diff
@@ -29,29 +32,47 @@ pub mod v1 {
...
-        });
-        STORAGE_VERSION.put::<Pallet<T>>();
-        let count = StoredJobRegistration::<T>::iter().count() as u64;
-        T::DbWeight::get().reads_writes(count + 1, count + 1)
-    } else {
-        Weight::zero()
+    let mut weight: Weight = Default::default();
+    for (i, f) in migrations.iter().enumerate() {
+        // correct index since we start with version 2 but i is 0-based
+        if onchain_version < StorageVersion::new((i + 2).try_into().unwrap()) {
```
Comment (RomarQ) [inline pallets/acurast/src/migration.rs:43]: What if we drop migration_v2?

File: pallets/assets-manager/src/lib.rs [hunk 3]
```diff
@@ -33,7 +33,8 @@ pub mod pallet {
     use xcm::prelude::{Abstract, AssetId, Concrete, GeneralIndex, PalletInstance, Parachain, X3};
     use xcm_executor::traits::Convert;
 
-    pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion::new(2);
+    pub(crate) const STORAGE_VERSION_: usize = 2;
+    pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion::new(STORAGE_VERSION_ as u16);
```
Comment (RomarQ) [inline pallets/assets-manager/src/lib.rs:37]: This pallet is going to be removed, right?

File: pallets/marketplace/src/migration.rs [hunk 4]
```diff
@@ -25,46 +28,56 @@ pub mod v1 {
...
-        let count = StoredAdvertisementRestriction::<T>::iter_values().count() as u64;
-        T::DbWeight::get().reads_writes(count + 1, count + 1)
-    } else {
-        Weight::zero()
+    let mut weight: Weight = Default::default();
+    for (i, f) in migrations.iter().enumerate() {
+        // correct index since we start with version 2 but i is 0-based
+        if onchain_version < StorageVersion::new((i + 2).try_into().unwrap()) {
+            weight += f();
+        }
```
Comment (RomarQ) [inline pallets/marketplace/src/migration.rs:44]: Same comments as in `acurast` pallet.

File: pallets/marketplace/src/migration.rs [hunk 5]
```diff
@@ -25,46 +28,56 @@ pub mod v1 {
     }
 }
 
-pub fn migrate_to_v2<T: Config>() -> Weight {
+pub fn migrate<T: Config>() -> Weight {
+    let migrations: [&dyn Fn() -> Weight; STORAGE_VERSION_ - 1] = [
```
Comment (RomarQ) [inline pallets/marketplace/src/migration.rs:32]: Same comments as in `acurast` pallet.

**Review by RomarQ — CHANGES_REQUESTED** [review]

File: pallets/assets-manager/src/lib.rs (hunk 3 above)
Comment (gitsimon) [reply pallets/assets-manager/src/lib.rs]: doesn't hurt to remove wrong code

**Comment by gitsimon** [comment]
@RomarQ simplified with explicit version numbers in migration array as discussed

**Review by RomarQ — APPROVED** [review]

**Pull request merged**